File: core/watchfit.go
// Fit the watch-time response curve per instance.
//
// Mirrors curator/model/watchfit.py exactly: the same fixed IRLS iteration
// count, the same ridge, the same Gaussian elimination with partial pivoting,
// and the same deterministic fold assignment by position. No RNG and no linear
// algebra library on either side, so the two implementations agree bit-for-bit
// the way the multi-hop PageRank kernel does.

package main

import "math"

var defaultViewCurve = [3]float64{-3.332, 0.734, -0.092}

const (
	minFirstPlays       = 200
	minPositives        = 30
	shrinkPriorStrength = 400.0
	payloadDigits       = 6
	cvFolds             = 5
	irlsIterations      = 40
	irlsRidge           = 1e-8
)

type viewCurveFit struct {
	coefficients [3]float64
	// logit of the instance's overall return rate; the curve is read against it.
	baseLogit        float64
	adopted          bool
	reason           string
	sampleSize       int
	positives        int
	heldoutQuadratic float64
	heldoutMonotone  float64
	heldoutConstant  float64
}

type watchSample struct {
	logT  float64
	label float64
}

func watchDesign(logT float64, terms int) [3]float64 {
	switch terms {
	case 1:
		return [3]float64{1, 0, 0}
	case 2:
		return [3]float64{1, logT, 0}
	default:
		return [3]float64{1, logT, logT * logT}
	}
}

// watchSolve mirrors watchfit._solve.
func watchSolve(matrix [][]float64, vector []float64) ([]float64, bool) {
	size := len(vector)
	augmented := make([][]float64, size)
	for row := 0; row < size; row++ {
		augmented[row] = make([]float64, size+1)
		copy(augmented[row], matrix[row])
		augmented[row][size] = vector[row]
	}
	for column := 0; column < size; column++ {
		pivotRow := column
		largest := math.Abs(augmented[column][column])
		for row := column + 1; row < size; row++ {
			candidate := math.Abs(augmented[row][column])
			if candidate > largest {
				largest = candidate
				pivotRow = row
			}
		}
		if largest < 1e-12 {
			return nil, false
		}
		if pivotRow != column {
			augmented[column], augmented[pivotRow] = augmented[pivotRow], augmented[column]
		}
		pivot := augmented[column][column]
		for row := column + 1; row < size; row++ {
			factor := augmented[row][column] / pivot
			if factor == 0 {
				continue
			}
			for col := column; col <= size; col++ {
				augmented[row][col] -= float64(factor * augmented[column][col])
			}
		}
	}
	solution := make([]float64, size)
	for row := size - 1; row >= 0; row-- {
		total := augmented[row][size]
		for col := row + 1; col < size; col++ {
			total -= float64(augmented[row][col] * solution[col])
		}
		solution[row] = total / augmented[row][row]
	}
	return solution, true
}

// watchFitLogistic mirrors watchfit._fit_logistic.
func watchFitLogistic(samples []watchSample, terms int) ([]float64, bool) {
	beta := make([]float64, terms)
	for iteration := 0; iteration < irlsIterations; iteration++ {
		hessian := make([][]float64, terms)
		for r := 0; r < terms; r++ {
			hessian[r] = make([]float64, terms)
			hessian[r][r] = irlsRidge
		}
		gradient := make([]float64, terms)
		for _, sample := range samples {
			row := watchDesign(sample.logT, terms)
			z := 0.0
			for index := 0; index < terms; index++ {
				z += float64(beta[index] * row[index])
			}
			p := 0.0
			if z > -700.0 {
				p = 1.0 / (1.0 + math.Exp(-z))
			}
			weight := p * (1.0 - p)
			if weight < 1e-9 {
				weight = 1e-9
			}
			residual := sample.label - p
			for r := 0; r < terms; r++ {
				gradient[r] += float64(row[r] * residual)
				for c := 0; c < terms; c++ {
					hessian[r][c] += float64(float64(row[r]*row[c]) * weight)
				}
			}
		}
		step, ok := watchSolve(hessian, gradient)
		if !ok {
			return nil, false
		}
		for index := 0; index < terms; index++ {
			beta[index] += step[index]
		}
		for index := 0; index < terms; index++ {
			if math.IsNaN(beta[index]) || math.IsInf(beta[index], 0) {
				return nil, false
			}
		}
	}
	return beta, true
}

// watchLogLikelihood mirrors watchfit._log_likelihood.
func watchLogLikelihood(samples []watchSample, beta []float64) float64 {
	terms := len(beta)
	total := 0.0
	for _, sample := range samples {
		row := watchDesign(sample.logT, terms)
		z := 0.0
		for index := 0; index < terms; index++ {
			z += float64(beta[index] * row[index])
		}
		if z > 0.0 {
			total += float64(sample.label*z) - (z + math.Log1p(math.Exp(-z)))
		} else {
			total += float64(sample.label*z) - math.Log1p(math.Exp(z))
		}
	}
	return total
}

// watchHeldout mirrors watchfit._heldout.
func watchHeldout(samples []watchSample, terms int) float64 {
	total := 0.0
	for fold := 0; fold < cvFolds; fold++ {
		train := make([]watchSample, 0, len(samples))
		test := make([]watchSample, 0, len(samples))
		for index, sample := range samples {
			if index%cvFolds != fold {
				train = append(train, sample)
			} else {
				test = append(test, sample)
			}
		}
		if len(train) == 0 || len(test) == 0 {
			continue
		}
		beta, ok := watchFitLogistic(train, terms)
		if !ok {
			return math.Inf(1)
		}
		total -= watchLogLikelihood(test, beta)
	}
	return total
}

// fitViewCurve mirrors watchfit.fit_view_curve. firstPlays must arrive in the
// same stable order the Python side uses (ordered by scene_id).
func fitViewCurve(firstPlays []struct {
	Seconds  float64
	Returned bool
}) viewCurveFit {
	samples := make([]watchSample, 0, len(firstPlays))
	positives := 0
	for _, play := range firstPlays {
		if play.Seconds <= 0 || math.IsNaN(play.Seconds) || math.IsInf(play.Seconds, 0) {
			continue
		}
		label := 0.0
		if play.Returned {
			label = 1.0
			positives++
		}
		samples = append(samples, watchSample{math.Log(play.Seconds), label})
	}
	inf := math.Inf(1)
	baseLogit := 0.0
	if positives > 0 && positives < len(samples) {
		baseRate := float64(positives) / float64(len(samples))
		baseLogit = math.Log(baseRate / (1.0 - baseRate))
	}
	refuse := func(reason string, quad, mono, konst float64) viewCurveFit {
		return viewCurveFit{
			defaultViewCurve, baseLogit, false, reason, len(samples), positives, quad, mono, konst,
		}
	}
	if len(samples) < minFirstPlays {
		return refuse("insufficient_sample", inf, inf, inf)
	}
	if positives < minPositives {
		return refuse("insufficient_positives", inf, inf, inf)
	}
	if positives == len(samples) {
		return refuse("no_negative_class", inf, inf, inf)
	}
	quadratic := watchHeldout(samples, 3)
	monotone := watchHeldout(samples, 2)
	constant := watchHeldout(samples, 1)
	if math.IsInf(quadratic, 0) || math.IsNaN(quadratic) {
		return refuse("fit_failed", quadratic, monotone, constant)
	}
	if quadratic >= monotone || quadratic >= constant {
		return refuse("not_better_than_baseline", quadratic, monotone, constant)
	}
	beta, ok := watchFitLogistic(samples, 3)
	if !ok {
		return refuse("fit_failed", quadratic, monotone, constant)
	}
	if beta[2] >= 0.0 {
		return refuse("curvature_not_negative", quadratic, monotone, constant)
	}
	weight := float64(len(samples)) / (float64(len(samples)) + shrinkPriorStrength)
	var shrunk [3]float64
	for index := 0; index < 3; index++ {
		shrunk[index] = float64(weight*beta[index]) + float64((1.0-weight)*defaultViewCurve[index])
	}
	if shrunk[2] >= 0.0 {
		return refuse("curvature_not_negative", quadratic, monotone, constant)
	}
	return viewCurveFit{
		shrunk, baseLogit, true, "adopted", len(samples), positives, quadratic, monotone, constant,
	}
}

// unfittedViewCurve is what the paths that do not fit use: the shipped curve,
// matching Python, where curation.py constructs a fresh PreferenceModelBuilder
// and calls _scene_labels() without a build having fitted anything.
func unfittedViewCurve() viewCurveFit {
	inf := math.Inf(1)
	return viewCurveFit{defaultViewCurve, 0, false, "not_fitted", 0, 0, inf, inf, inf}
}

// curve is the four numbers viewingOutcomeCurve needs: the fit plus its centre.
func (fit viewCurveFit) curve() [4]float64 {
	return [4]float64{fit.coefficients[0], fit.coefficients[1], fit.coefficients[2], fit.baseLogit}
}

// payload mirrors ViewCurveFit.as_payload, including the null-for-infinite
// rule: json.dumps would emit Infinity, which is not JSON.
func (fit viewCurveFit) payload() jVal {
	scale := math.Pow(10, payloadDigits)
	round := func(value float64) float64 {
		return math.Round(value*scale) / scale
	}
	score := func(value float64) jVal {
		if math.IsInf(value, 0) || math.IsNaN(value) {
			return jvNull()
		}
		return jvFloat(round(value))
	}
	coefficients := jvArr(
		jvFloat(round(fit.coefficients[0])),
		jvFloat(round(fit.coefficients[1])),
		jvFloat(round(fit.coefficients[2])),
	)
	return jvObj(
		jvKey("coefficients", coefficients),
		jvKey("base_logit", jvFloat(round(fit.baseLogit))),
		jvKey("adopted", jvBool(fit.adopted)),
		jvKey("reason", jvStr(fit.reason)),
		jvKey("sample_size", jvInt(int64(fit.sampleSize))),
		jvKey("positives", jvInt(int64(fit.positives))),
		jvKey("heldout_quadratic", score(fit.heldoutQuadratic)),
		jvKey("heldout_monotone", score(fit.heldoutMonotone)),
		jvKey("heldout_constant", score(fit.heldoutConstant)),
	)
}

File: curator/model/watchfit.py
"""Fit the watch-time response curve per instance.

`viewing_outcome`'s shipped curve rises monotonically in watch time. Measured
against an outcome it cannot influence -- whether the user returned to the
scene on a later day -- the real relationship is an inverted U peaking near a
minute, so the shipped curve assigns its maximum where measured outcomes are
near-worst.

This fits `logit(p) = b0 + b1*ln t + b2*(ln t)^2` on first plays and returns
coefficients only when they earn adoption. Everything here is deliberately
plain float arithmetic in a fixed operation order, with no RNG and no linear
algebra library: the Go model builder mirrors it bit-for-bit, the same way the
multi-hop PageRank kernel is mirrored.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

# Shipped prior. These are the coefficients measured on a real library, and
# they are what a fitted curve is shrunk toward, so a small instance moves
# smoothly from shipped to personal behaviour instead of jumping between
# builds.
DEFAULT_VIEW_CURVE = (-3.332, 0.734, -0.092)

# Guards. A cold install must never be worse off than the shipped curve.
MIN_FIRST_PLAYS = 200
MIN_POSITIVES = 30
SHRINK_PRIOR_STRENGTH = 400.0
PAYLOAD_DIGITS = 6
CV_FOLDS = 5
IRLS_ITERATIONS = 40
IRLS_RIDGE = 1e-8


@dataclass(frozen=True)
class ViewCurveFit:
    """The outcome of a fit attempt, including why it was refused."""

    coefficients: tuple[float, float, float]
    # logit of the instance's overall return rate. The curve is read relative
    # to this: a duration that predicts returning better than the library
    # average is positive evidence, one that predicts worse is negative.
    base_logit: float
    adopted: bool
    reason: str
    sample_size: int
    positives: int
    heldout_quadratic: float
    heldout_monotone: float
    heldout_constant: float

    @property
    def curve(self) -> tuple[float, float, float, float]:
        """The four numbers `viewing_outcome` needs: the fit plus its centre."""
        return (*self.coefficients, self.base_logit)

    def as_payload(self) -> dict[str, object]:
        """Recorded in the artifact so a model says which curve produced it.

        Held-out scores are null rather than infinite when a guard refused
        before scoring: `json.dumps` would emit `Infinity`, which is not JSON
        and which the Go mirror could not reproduce.

        Values are rounded because Go's math.Exp/math.Log and CPython's libm
        differ in the last bit, and this payload is compared as a JSON string
        between the two implementations. Six digits is far more than the field
        needs -- it is provenance, not an input to anything -- and far coarser
        than the disagreement, so the two always serialize identically.
        """

        def score(value: float) -> float | None:
            return round(value, PAYLOAD_DIGITS) if math.isfinite(value) else None

        return {
            "coefficients": [round(value, PAYLOAD_DIGITS) for value in self.coefficients],
            "base_logit": round(self.base_logit, PAYLOAD_DIGITS),
            "adopted": self.adopted,
            "reason": self.reason,
            "sample_size": self.sample_size,
            "positives": self.positives,
            "heldout_quadratic": score(self.heldout_quadratic),
            "heldout_monotone": score(self.heldout_monotone),
            "heldout_constant": score(self.heldout_constant),
        }


def _design(log_t: float, terms: int) -> tuple[float, ...]:
    if terms == 1:
        return (1.0,)
    if terms == 2:
        return (1.0, log_t)
    return (1.0, log_t, log_t * log_t)


def _solve(matrix: list[list[float]], vector: list[float]) -> list[float] | None:
    """Gaussian elimination with partial pivoting on a small dense system.

    Written out rather than delegated so the Go mirror can reproduce the same
    operation order exactly.
    """
    size = len(vector)
    augmented = [[*matrix[row], vector[row]] for row in range(size)]
    for column in range(size):
        pivot_row = column
        largest = abs(augmented[column][column])
        for row in range(column + 1, size):
            candidate = abs(augmented[row][column])
            if candidate > largest:
                largest = candidate
                pivot_row = row
        if largest < 1e-12:
            return None
        if pivot_row != column:
            augmented[column], augmented[pivot_row] = augmented[pivot_row], augmented[column]
        pivot = augmented[column][column]
        for row in range(column + 1, size):
            factor = augmented[row][column] / pivot
            if factor == 0.0:
                continue
            for col in range(column, size + 1):
                augmented[row][col] -= factor * augmented[column][col]
    solution = [0.0] * size
    for row in range(size - 1, -1, -1):
        total = augmented[row][size]
        for col in range(row + 1, size):
            total -= augmented[row][col] * solution[col]
        solution[row] = total / augmented[row][row]
    return solution


def _fit_logistic(samples: list[tuple[float, float]], terms: int) -> list[float] | None:
    """Newton/IRLS on the logistic likelihood. Fixed iteration count so two
    implementations cannot diverge on a convergence test."""
    beta = [0.0] * terms
    for _ in range(IRLS_ITERATIONS):
        hessian = [[IRLS_RIDGE if r == c else 0.0 for c in range(terms)] for r in range(terms)]
        gradient = [0.0] * terms
        for log_t, label in samples:
            row = _design(log_t, terms)
            z = 0.0
            for index in range(terms):
                z += beta[index] * row[index]
            p = 1.0 / (1.0 + math.exp(-z)) if z > -700.0 else 0.0
            weight = p * (1.0 - p)
            if weight < 1e-9:
                weight = 1e-9
            residual = label - p
            for r in range(terms):
                gradient[r] += row[r] * residual
                for c in range(terms):
                    hessian[r][c] += row[r] * row[c] * weight
        step = _solve(hessian, gradient)
        if step is None:
            return None
        for index in range(terms):
            beta[index] += step[index]
        if not all(math.isfinite(value) for value in beta):
            return None
    return beta


def _log_likelihood(samples: list[tuple[float, float]], beta: list[float]) -> float:
    terms = len(beta)
    total = 0.0
    for log_t, label in samples:
        row = _design(log_t, terms)
        z = 0.0
        for index in range(terms):
            z += beta[index] * row[index]
        # log(1 + e^z) without overflow.
        total += label * z - (z + math.log1p(math.exp(-z)) if z > 0.0 else math.log1p(math.exp(z)))
    return total


def _heldout(samples: list[tuple[float, float]], terms: int) -> float:
    """Negative held-out log-likelihood over deterministic folds.

    Folds are assigned by position in the caller's stable ordering, so no
    random seed has to agree across two languages.
    """
    total = 0.0
    for fold in range(CV_FOLDS):
        train = [row for index, row in enumerate(samples) if index % CV_FOLDS != fold]
        test = [row for index, row in enumerate(samples) if index % CV_FOLDS == fold]
        if not train or not test:
            continue
        beta = _fit_logistic(train, terms)
        if beta is None:
            return math.inf
        total -= _log_likelihood(test, beta)
    return total


def fit_view_curve(first_plays: list[tuple[float, bool]]) -> ViewCurveFit:
    """Fit and adjudicate the curve. `first_plays` must be in a stable order."""
    samples = [
        (math.log(seconds), 1.0 if returned else 0.0)
        for seconds, returned in first_plays
        if seconds > 0.0 and math.isfinite(seconds)
    ]
    positives = sum(1 for _, label in samples if label > 0.0)
    base_logit = 0.0
    if 0 < positives < len(samples):
        base_rate = positives / len(samples)
        base_logit = math.log(base_rate / (1.0 - base_rate))

    def refuse(
        reason: str, quad: float = math.inf, mono: float = math.inf, const: float = math.inf
    ) -> ViewCurveFit:
        return ViewCurveFit(
            DEFAULT_VIEW_CURVE,
            base_logit,
            False,
            reason,
            len(samples),
            positives,
            quad,
            mono,
            const,
        )

    if len(samples) < MIN_FIRST_PLAYS:
        return refuse("insufficient_sample")
    if positives < MIN_POSITIVES:
        return refuse("insufficient_positives")
    if positives == len(samples):
        # Every first play returned: there is no base rate to read the curve
        # against, so there is nothing for it to be evidence relative to.
        return refuse("no_negative_class")

    quadratic = _heldout(samples, 3)
    monotone = _heldout(samples, 2)
    constant = _heldout(samples, 1)
    if not math.isfinite(quadratic):
        return refuse("fit_failed", quadratic, monotone, constant)
    # "It won", not "it converged": the shape has to beat both simpler
    # alternatives on data it did not see.
    if quadratic >= monotone or quadratic >= constant:
        return refuse("not_better_than_baseline", quadratic, monotone, constant)

    beta = _fit_logistic(samples, 3)
    if beta is None:
        return refuse("fit_failed", quadratic, monotone, constant)
    if beta[2] >= 0.0:
        # Positive curvature is a U, not an inverted U -- the shipped curve is
        # a better description than that.
        return refuse("curvature_not_negative", quadratic, monotone, constant)

    weight = len(samples) / (len(samples) + SHRINK_PRIOR_STRENGTH)
    shrunk = tuple(
        weight * beta[index] + (1.0 - weight) * DEFAULT_VIEW_CURVE[index] for index in range(3)
    )
    if shrunk[2] >= 0.0:
        return refuse("curvature_not_negative", quadratic, monotone, constant)
    return ViewCurveFit(
        (shrunk[0], shrunk[1], shrunk[2]),
        base_logit,
        True,
        "adopted",
        len(samples),
        positives,
        quadratic,
        monotone,
        constant,
    )
